packages/opentelemetry-semantic-conventions/src/internal/constants.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export const ACCOUNT_DOT = 'account.';
```

Comment:
>). It also "constructs" the static strings at runtime rather than including all of them, to aid in minification of the bundle .

Ever since you mentioned this in SIGs, I've been interested to see how well it compares to just gzip/brotli, as it's pretty much the same thing they do (find recurring parts and de-duplicate them) algorithmically & automatically. So I ran some tests for it

---

For context for everyone else less interested in the topic, you can pretty much split code size impact into 3 categories:

1. How long js takes to download (impacted by bandwidth - from fiber to 2G or EDGE),
2. how long it takes to parse (impacted by how fast CPU - be it from latest overclocked CPU to mobile chip)
3. and how long it takes to evaluate (same as above)

Note that for the first one, it's reasonable to expect the servers are gonna serve gzip/brotli'd file as it's a best practise that's in every production config and is-your-site-fast tool check. 2 & 3 happen on decompressed code so raw file size is more impactful, though there might be a case made for adding two strings together taking insignificantly (really, all else equal it doesn't matter) more work

---

So I grabbed this PR's code, compiled it, took esm files to give the best tree-shake chance, made a copy where I then undid all of this work, and installed [size-limit](https://github.com/ai/size-limit) to measure it - it bundles via webpack/esbuild, reports filesize (min / gzip / brotli) and measures how long it takes via headless chrome. For maximum compression benefit (maximum repeating strings = best compression case for both manual and automatic), I did an import of every exported string value

Here's the repository for reference & reproducing: https://github.com/t2t2/split-strings-test

First file size test. Ran it with webpack, then swapped for esbuild as size-limit supports both:

```
npm run size
npm uninstall @size-limit/webpack
npm install --save-dev @size-limit/esbuild
npm run size
```

| | Split strings (as PR) | Merged strings (as prev) | PR compared to merged
---|---|---|---
Webpack - size (min) | 7.36 kB | 7.99 kB | 92.1% 🔽
Webpack - size (gzip) | 3.57 kB | 3.25 kB | 109.8% 🔺
Webpack - size (brotli) | 3.1 kB | 2.74 kB | 113.1% 🔺
esbuild - size (min) | 8.33 kB | 9.04 kB | 92.1% 🔽
esbuild - size (gzip) | 4.23 kB | 3.97 kB | 106.5% 🔺
esbuild - size (brotli) | 3.5 kB | 3.17 kB | 110.4% 🔺

```
✔ Adding to empty webpack project
✔ Running JS in headless Chrome

  split strings
  Size:         3.1 kB with all dependencies, minified and brotlied
  Loading time: 61 ms  on slow 3G
  Running time: 151 ms on Snapdragon 410
  Total time:   211 ms

  merged strings
  Size:         2.74 kB with all dependencies, minified and brotlied
  Running time: 159 ms  on Snapdragon 410
  Total time:   212 ms

  merged strings (run 2)
  Size:         2.74 kB with all dependencies, minified and brotlied
  Loading time: 54 ms   on slow 3G
  Running time: 142 ms  on Snapdragon 410
  Total time:   195 ms

  split strings (run 2)
  Size:         3.1 kB with all dependencies, minified and brotlied
  Loading time: 61 ms  on slow 3G
  Running time: 160 ms on Snapdragon 410
  Total time:   220 ms
```

So according to data, while this would decrease the size of the raw file by ~8%, **it would actually increase the size of compressed file by around 10%** depending on compression algorithm - which would actually impact download size

---

When it comes to comparing how long these take to parse and evaluate, I ran size-limit multiple times and couldn't get reliably consistent results between runs. I think because it's comparing such short durations, the overhead & noise from background processes makes it too hard to measure

<details>
<summary>Here's the data I got, it's just too varying for me to base any conclusions on</summary>

|  | Split strings (as PR) | Merged strings (as one)
---|---|---
Run 1.1 | 151 ms | 159 ms
Run 1.2 | 160 ms | 142 ms
Run 2.1 | 155 ms | 117 ms
Run 2.2 | 153 ms | 164 ms
Run 3.1 | 138 ms | 123 ms
Run 3.2 | 132 ms | 153 ms
Run 4.1 | 152 ms | 149 ms
Run 4.2 | 198 ms | 134 ms
Run 5.1 | 170 ms | 159 ms
Run 5.2 | 149 ms | 164 ms
</details>

---

So as far as I can measure, this approach doesn't seem to have any actual minification wins, especially when it comes to downloaded bundle size from your average server serving static, compressed files

Reply:
> One aspect that is not highlighted is how this aids in tree-shaking, as this will enable (once everything stops importing / using the current namespaced "class") ...

This isn't what I'm talking about, you may note from the test repo that for both cases I imported the individual string exports ([list of imports](https://github.com/t2t2/split-strings-test/blob/main/.size-limit.cjs#L1), [generator func](https://github.com/t2t2/split-strings-test/blob/main/get-exports.js)) not the objects (which yep objects can't be treeshaken, that's bad js, no disagreement here); so both are equally tree-shakeable. In fact, I've added [two more test cases, importing only HTTP in one and only NET in other bundle](https://github.com/t2t2/split-strings-test/blob/main/.size-limit.cjs#L3):

For the first one:

```js
import '{ SEMATTRS_HTTP_CLIENT_IP, SEMATTRS_HTTP_FLAVOR, SEMATTRS_HTTP_HOST, SEMATTRS_HTTP_METHOD, SEMATTRS_HTTP_REQUEST_CONTENT_LENGTH, SEMATTRS_HTTP_REQUEST_CONTENT_LENGTH_UNCOMPRESSED, SEMATTRS_HTTP_RESPONSE_CONTENT_LENGTH, SEMATTRS_HTTP_RESPONSE_CONTENT_LENGTH_UNCOMPRESSED, SEMATTRS_HTTP_ROUTE, SEMATTRS_HTTP_SCHEME, SEMATTRS_HTTP_SERVER_NAME, SEMATTRS_HTTP_STATUS_CODE, SEMATTRS_HTTP_TARGET, SEMATTRS_HTTP_URL, SEMATTRS_HTTP_USER_AGENT }' from './[path].js;
```

| | Split strings (as PR) | Merged strings (as prev) | PR compared to merged
---|---|---|---
Webpack - size (min) | 3.91 kB | 4.72 kB | -0.81kB
Webpack - size (gzip) | 1.81 kB | 1.52 kB | +0.29kB
Webpack - size (brotli) | 1.6 kB | 1.3 kB | +0.3kB

<details><summary>Bundle output: Split strings</summary>

```js
(()=>{"use strict";var e=function(e){for(var t={},s=e.length,o=0;o<s;o++){var n=e[o];n&&(t[String(n).toUpperCase().replace(/[-.]/g,"_")]=n)}return t},t="arn",s="aws",o="cassandra.",n="carrier.",a="cloud.",r="cluster.",i="code.",c="collection",d="command",l="compressed",p="container.",m="content",_="connection",u="coordinator.",g="cron",b="daemon",y="db.",v="deployment.",f="description",h="device.",k="document.",x=".",w="dynamodb.",z="ecs.",j="executable.",q="enduser.",C="exception.",S="faas.",U="global",A="group.",B="kafka.",D="k8s.",E="http.",F="host",G="id",H="image.",I="index",J="instance",K="invoked",L="job",M="jsonrpc.",N="length",O="log.",P="message",Q="messaging.",R="model.",T="name",V="net.",W="node.",X="os.",Y="peer.",Z="port",$="process.",ee="protocol",te="provisioned",se="request",oe="replica",ne="response",ae="rpc.",re="runtime.",ie="sdk.",ce="secondary",de="service.",le="set.",pe="space",me="stateful",_e="stream.",ue="task.",ge="telemetry.",be="thread.",ye="type",ve="uid",fe="un",he="_",ke="webengine.",xe="version",we=E+"method",ze=E+"url",je=E+"target",qe=E+F,Ce=E+"scheme",Se=E+"status_code",Ue=E+"flavor",Ae=E+"user_agent",Be=E+se+he+m+he+N,De=E+se+he+m+he+N+he+fe+l,Ee=E+ne+he+m+he+N,Fe=E+ne+he+m+he+N+he+fe+l,Ge=E+"server_name",He=E+"route",Ie=E+"client_ip";e(["aws.lambda."+K+"_arn",y+"system",y+_+he+"string",y+"user",y+"jdbc.driver_classname",y+T,y+"statement",y+"operation",y+"mssql."+J+he+T,y+o+"keyspace",y+o+"page_size",y+o+"consistency_level",y+o+"table",y+o+G+"empotence",y+o+"speculative_execution_count",y+o+u+G,y+o+u+"dc",y+"hbase.namespace",y+"redis.database_index",y+"mongodb."+c,y+"sql.table",C+ye,C+P,C+"stacktrace",C+"escaped",S+"trigger",S+"execution",S+k+c,S+k+"operation",S+k+"time",S+k+T,S+"time",S+"cron",S+"coldstart",S+K+he+T,S+K+he+"provider",S+K+he+"region",V+"transport",V+Y+"ip",V+Y+Z,V+Y+T,V+F+x+"ip",V+F+x+Z,V+F+x+T,V+F+x+_+x+ye,V+F+x+_+x+"subtype",V+F+x+n+T,V+F+x+n+"mcc",V+F+x+n+"mnc",V+F+x+n+"icc",Y+"service",q+G,q+"role",q+"scope",be+G,be+T,i+"function",i+T+pe,i+"filepath",i+"lineno",we,ze,je,qe,Ce,Se,Ue,Ae,Be,De,Ee,Fe,Ge,He,Ie,s+x+w+"table_names",s+x+w+"consumed_capacity",s+x+w+"item_collection_metrics",s+x+w+te+he+"read_capacity",s+x+w+te+he+"write_capacity",s+x+w+"consistent_read",s+x+w+"projection",s+x+w+"limit",s+x+w+"attributes_to_get",s+x+w+I+he+T,s+x+w+"select",s+x+w+U+he+ce+he+I+"es",s+x+w+"local"+he+ce+he+I+"es",s+x+w+"exclusive_start_table",s+x+w+"table_count",s+x+w+"scan_forward",s+x+w+"segment",s+x+w+"total_segments",s+x+w+"count",s+x+w+"scanned_count",s+x+w+"attribute_definitions",s+x+w+U+he+ce+he+I+he+"updates",Q+"system",Q+"destination",Q+"destination_kind",Q+"temp_destination",Q+ee,Q+ee+he+xe,Q+"url",Q+P+he+G,Q+"conversation_id",Q+P+he+"payload_size_bytes",Q+P+he+"payload_compressed_size_bytes",Q+"operation",Q+"consumer_id",Q+"rabbitmq.routing_key",Q+B+P+he+"key",Q+B+"consumer_group",Q+B+"client_id",Q+B+"partition",Q+B+"tombstone",ae+"system",ae+"service",ae+"method",ae+"grpc.status_code",ae+M+xe,ae+M+se+he+G,ae+M+"error_code",ae+M+"error_message",P+x+ye,P+x+G,P+x+l+he+"size",P+x+fe+l+he+"size"]),e([a+"provider",a+"account."+G,a+"region",a+"availability_zone",a+"platform",s+x+z+p+t,s+x+z+r+t,s+x+z+"launchtype",s+x+z+ue+t,s+x+z+ue+"family",s+x+z+ue+"revision","aws.eks."+r+t,s+x+O+A+T+"s",s+x+O+A+t+"s",s+x+O+_e+T+"s",s+x+O+_e+t+"s",p+T,p+G,p+"runtime",p+H+T,p+H+"tag",v+"environment",h+G,h+R+G+"entifier",h+R+T,S+T,S+G,S+xe,S+J,S+"max_memory",F+x+G,F+x+T,F+x+ye,F+x+"arch",F+x+H+T,F+x+H+G,F+x+H+xe,D+r+T,D+W+T,D+W+ve,D+T+pe+x+T,D+"pod.uid",D+"pod.name",D+p+T,D+oe+le+ve,D+oe+le+T,D+v+ve,D+v+T,D+me+le+ve,D+me+le+T,D+b+le+ve,D+b+le+T,D+L+x+ve,D+L+x+T,D+g+L+x+ve,D+g+L+x+T,X+ye,X+f,X+T,X+xe,$+"pid",$+j+T,$+j+"path",$+d,$+d+he+"line",$+d+he+"args",$+"owner",$+re+T,$+re+xe,$+re+f,de+T,de+T+pe,de+J+x+G,de+xe,ge+ie+T,ge+ie+"language",ge+ie+xe,ge+"auto.version",ke+T,ke+xe,ke+f]),console.log(Ie,Ue,qe,we,Be,De,Ee,Fe,He,Ce,Ge,Se,je,ze,Ae)})();
```
</details>
<details><summary>Build output: Merged strings</summary>

```js
(()=>{"use strict";var e=function(e){for(var s={},a=e.length,n=0;n<a;n++){var t=e[n];t&&(s[String(t).toUpperCase().replace(/[-.]/g,"_")]=t)}return s},s="http.method",a="http.url",n="http.target",t="http.host",o="http.scheme",i="http.status_code",r="http.flavor",d="http.user_agent",c="http.request_content_length",m="http.request_content_length_uncompressed",p="http.response_content_length",g="http.response_content_length_uncompressed",l="http.server_name",_="http.route",b="http.client_ip";e(["awslambda.invoked_arn","db.system","db.connection_string","db.user","db.jdbc.driver_classname","db.name","db.statement","db.operation","db.mssql.instance_name","db.cassandra.keyspace","db.cassandra.page_size","db.cassandra.consistency_level","db.cassandra.table","db.cassandra.idempotence","db.cassandra.speculative_execution_count","db.cassandra.coordinator.id","db.cassandra.coordinator.dc","db.hbase.namespace","db.redis.database_index","db.mongodb.collection","db.sql.table","exception.type","exception.message","exception.stacktrace","exception.escaped","faas.trigger","faas.execution","faas.document.collection","faas.document.operation","faas.document.time","faas.document.name","faas.time","faas.cron","faas.coldstart","faas.invoked_name","faas.invoked_provider","faas.invoked_region","net.transport","net.peer.ip","net.peer.port","net.peer.name","net.hostip","net.hostport","net.hostname","net.hostconnectiontype","net.hostconnectionsubtype","net.hostcarrier.name","net.hostcarrier.mcc","net.hostcarrier.mnc","net.hostcarrier.icc","peer.service","enduser.id","enduser.role","enduser.scope","thread.id","thread.name","code.function","code.namespace","code.filepath","code.lineno",s,a,n,t,o,i,r,d,c,m,p,g,l,_,b,"awsdynamodb.table_names","awsdynamodb.consumed_capacity","awsdynamodb.item_collection_metrics","awsdynamodb.provisioned_read_capacity","awsdynamodb.provisioned_write_capacity","awsdynamodb.consistent_read","awsdynamodb.projection","awsdynamodb.limit","awsdynamodb.attributes_to_get","awsdynamodb.index_name","awsdynamodb.select","awsdynamodb.global_secondary_indexes","awsdynamodb.local_secondary_indexes","awsdynamodb.exclusive_start_table","awsdynamodb.table_count","awsdynamodb.scan_forward","awsdynamodb.segment","awsdynamodb.total_segments","awsdynamodb.count","awsdynamodb.scanned_count","awsdynamodb.attribute_definitions","awsdynamodb.global_secondary_index_updates","messaging.system","messaging.destination","messaging.destination_kind","messaging.temp_destination","messaging.protocol","messaging.protocol_version","messaging.url","messaging.message_id","messaging.conversation_id","messaging.message_payload_size_bytes","messaging.message_payload_compressed_size_bytes","messaging.operation","messaging.consumer_id","messaging.rabbitmq.routing_key","messaging.kafka.message_key","messaging.kafka.consumer_group","messaging.kafka.client_id","messaging.kafka.partition","messaging.kafka.tombstone","rpc.system","rpc.service","rpc.method","rpc.grpc.status_code","rpc.jsonrpc.version","rpc.jsonrpc.request_id","rpc.jsonrpc.error_code","rpc.jsonrpc.error_message","messagetype","messageid","messagecompressed_size","messageuncompressed_size"]),e(["cloud.provider","cloud.account.id","cloud.region","cloud.availability_zone","cloud.platform","awsecs.container.arn","awsecs.cluster.arn","awsecs.launchtype","awsecs.task.arn","awsecs.task.family","awsecs.task.revision","awseks.cluster.arn","awslog.group.names","awslog.group."+ARN+"s","awslog.stream.names","awslog.stream."+ARN+"s","container.name","container.id","container.runtime","container.image.name","container.image.tag","deployment.environment","device.id","device.model.identifier","device.model.name","faas.name","faas.id","faas.version","faas.instance","faas.max_memory","hostid","hostname","hosttype","hostarch","hostimage.name","hostimage.id","hostimage.version","k8s.cluster.name","k8s.node.name","k8s.node.uid","k8s.namespacename","k8s.pod.uid","k8s.pod.name","k8s.container.name","k8s.replicaset.uid","k8s.replicaset.name","k8s.deployment.uid","k8s.deployment.name","k8s.statefulset.uid","k8s.statefulset.name","k8s.daemonset.uid","k8s.daemonset.name","k8s.jobuid","k8s.jobname","k8s.cronjobuid","k8s.cronjobname","os.type","os.description","os.name","os.version","process.pid","process.executable.name","process.executable.path","process.command","process.command_line","process.command_args","process.owner","process.runtime.name","process.runtime.version","process.runtime.description","service.name","service.namespace","service.instanceid","service.version","telemetry.sdk.name","telemetry.sdk.language","telemetry.sdk.version","telemetry.auto.version","webengine.name","webengine.version","webengine.description"]),console.log(b,r,t,s,c,m,p,g,_,o,l,i,n,a,d)})();
```
</details>

(for anyone following along, [beautifier.io](https://beautifier.io/) is a great tool to read these)

Actually looking at this output, looks like you've forgotten to mark `createConstMap` [calls as pure so it didn't get tree-shaken](https://webpack.js.org/guides/tree-shaking/#mark-a-function-call-as-side-effect-free) (also why otel/api package doesn't get tree-shaken per signal). [Quick fix later to both versions](https://github.com/t2t2/split-strings-test/commit/fd0d55d0a436085be6e5e673eae52b9783f141e7):

| | Split strings (as PR) | Merged strings (as prev) | PR compared to merged
---|---|---|---
Webpack - size (min) | 340 B | 320 B | +20 B
Webpack - size (gzip) | 192 B | 128 B | +64 B
Webpack - size (brotli) | 179 B | 116 B | +63 B

<details><summary>Bundle output: Split strings</summary>

```js
(()=>{"use strict";var e="compressed",t="content",s="http.",r="length",o="request",n="response",u="_",c=s+"method",a=s+"url",l=s+"target",h=s+"host",_=s+"scheme",g=s+"status_code",m=s+"flavor",p=s+"user_agent",d=s+o+u+t+u+r,i=s+o+u+t+u+r+u+"un"+e,v=s+n+u+t+u+r,f=s+n+u+t+u+r+u+"un"+e,q=s+"server_name",b=s+"route",j=s+"client_ip";console.log(j,m,h,c,d,i,v,f,b,_,q,g,l,a,p)})();
```
</details>
<details><summary>Build output: Merged strings</summary>

```js
(()=>{"use strict";ARN,ARN,console.log("http.client_ip","http.flavor","http.host","http.method","http.request_content_length","http.request_content_length_uncompressed","http.response_content_length","http.response_content_length_uncompressed","http.route","http.scheme","http.server_name","http.status_code","http.target","http.url","http.user_agent")})();
```
</details>

Analysing the output the anomaly for minified size comes from minifier inlining variables that are only used once, while it's not done at all for the split variables. I guess add that as another W to keeping strings together


Anyways; Still, I'm importing specific strings from exports, running them through bundler+minifier, and with treeshaking fixed, the results I get still say that `d = s + o + u + t + u + r` gives worse compressed bundle size than `d = "http.request_content_length"`, which as any decent CDN uses compression it means this is the worse approach and you should just let gzip/brotli do it's thing

> This also helps with the final memory usage on the client as now there is also less "code" to be held in-memory (or jit'ed), although depending on the runtime the constructed "strings" can't directly reference the "code" embedded string (depends on whether the JS runtime implements this sort of optimization)

I look forward to any sort of proof about this, as in my mind during evaluation it would store strings in memory, then merge the strings together storing the new string in memory and eventually garbage collecting the parts as they're no longer referenced anywhere, eg:

1. set variable1 = "http."
2. set variable2 = "request"
3. ...
4. set variableN = "http.request_content_length" (combined from prev vars)
5. ...a bit later during garbage collection...
6. delete variables used to construct combined vars as they're no longer directly referenced

vs.

1. set variableN = "http.request_content_length"

Reply:
Nice tests.

I'll look deeper into what's going on here (it will most likely be next week before I get some real time into this) as looking at the output provided there are split strings that got re-combined in the output that should not have (which by eyeballing (may) reduce the un-minified by a few bytes (not 20)).

> the results I get still say that d = s + o + u + t + u + r gives worse compressed bundle size than d = "http.request_content_length", which as any decent CDN uses compression it means this is the worse approach

Well this depends on what is actually getting compressed and whether there are any duplicated sequences to get compressed.  And I agree that it has issues.

And I think it's because as part of using the template generation approach that I've used I, one of the primary ones being that it's not "really" accounting for each "substring" needing to be declared, which is a minimum of 9 bytes of overhead `var x="";` plus any specific usage scenario (2 bytes for just `+x` 3 for `x+"` so when it's only used once / twice it's a problem.

In my earlier testing I avoided having any constants there where less than 4 characters, but as I got closer to the published version I added the `.` and `_` variants which in your generated example have become `fe` and `he` along with `+` to glue things together), likewise for `un`, as adding these allowed the "full" version to become smaller because it could now reuse and "construct" more (longer) strings like `compressed` and `uncompressed` (maybe this is a mistake, if we look at the longer term vision (of fully tree-shaken), and having No "fully" (namespaced objects), in which case I would immediately (try) removing anything smaller than 4 characters and any other "string" that is not used more than 4 times. not-withstanding being able to create a post processor to do this automatically, part of the stucture that I put in place in the template was to be able to more "easily" update the generated strings as the semantic conventions change as this will affect the "manually" identified sub-strings.

But generally, overall yes, I did expect that the compressed form would not be as small as I was focusing (by eyeballing) on creating the smallest possible overall minified version (ie the fully namespaced version) as I suspect that it will be a long time before we fully remove all of the namespaced usages (which instantly drag in every string).

Webpack stuff
Needing to mark the usage of the function as `/*# PURE /*` for webpack is just annoying 😢, as we use rollup for our packages we don't see this issue (in our generated code).
One really bad webpack pattern (which I'm happy not to see in your example project) is when webpack "wraps" a generated constant strings into individual functions and all usages call the function so that the returned "value" can't be changed. (which IS technically a potential issue in pure JS -- ie. somehow change a base part of the name and it's now generating different strings) -- only ever saw this once, for an internal team so I don't know which versions / config they had set.

Reply:
While I can get the "split" string version of the bundles smaller (by 0.5 -> 1kb) than the full string and I managed a couple of combinations where the gzip was also smaller (but not the brotli) it was only single digit bytes.

So based on the complexity of the template for the very minimal improvement and that we need to go through a semantic convention version "update". I'm just going to simplify this PR to just split out the string without splitting.

I did find from testing that to get the "better" overall size and to stop the CommonJS and minified versions from referencing the unminifiable "exported" names it will include some local "temp" variables to assign to the single "Named" exports as well as for the namespaced objects, so that for the near term anyone referencing the namespace and with individual strings will at least reduce the duplication of the string etc.